src/adapter/src/catalog/open/builtin_item_migration.rs

```diff
@@ -0,0 +1,527 @@
+            } else {
+                migration_shards_to_finalize.insert((table_key, shard_id));
+            }
+        }
+    }
+
+    // 5. Add migrated tables to migration shard for current generation.
+    let mut global_id_shards: BTreeMap<_, _> = global_id_shards
+        .into_iter()
+        .map(|(table_key, shard_id)| (table_key.global_id, shard_id))
+        .collect();
+    for global_id in migrated_tables {
+        if !global_id_shards.contains_key(&global_id) {
+            let shard_id = ShardId::new();
+            global_id_shards.insert(global_id, shard_id);
+            let table_key = TableKey {
+                global_id,
+                deploy_generation,
+            };
+            migrated_shard_updates.push(((table_key, shard_id), upper, 1));
+        }
+    }
+
+    write_to_migration_shard(
+        migrated_shard_updates,
+        upper,
+        &mut write_handle,
+        &mut since_handle,
+    )
+    .await?;
+
+    // 6. Update `GlobalId` to `ShardId` mapping and finalize old `ShardId`s.
+    {
+        let txn: &mut dyn StorageTxn<Timestamp> = txn;
+        let storage_collection_metadata = txn.get_collection_metadata();
+        let global_id_shards: BTreeMap<_, _> = global_id_shards
+            .into_iter()
+            .map(|(global_id, shard_id)| (GlobalId::System(global_id), shard_id))
+            .filter(|(global_id, shard_id)| {
+                storage_collection_metadata.get(global_id) != Some(shard_id)
+            })
+            .collect();
+        let global_ids: BTreeSet<_> = global_id_shards.keys().cloned().collect();
+        let mut old_shard_ids: BTreeSet<_> = txn
+            .delete_collection_metadata(global_ids)
+            .into_iter()
+            .map(|(_, shard_id)| shard_id)
+            .collect();
+        old_shard_ids.extend(
+            migration_shards_to_finalize
+                .iter()
+                .map(|(_, shard_id)| shard_id),
+        );
+        txn.insert_unfinalized_shards(old_shard_ids).map_err(|e| {
+            Error::new(ErrorKind::Internal(format!(
+                "builtin migration failed: {e}"
+            )))
+        })?;
+        txn.insert_collection_metadata(global_id_shards)
+            .map_err(|e| {
+                Error::new(ErrorKind::Internal(format!(
+                    "builtin migration failed: {e}"
+                )))
+            })?;
+    }
+
+    let updates = txn.get_and_commit_op_updates();
+    let builtin_table_updates = state.apply_updates_for_bootstrap(updates).await;
+
+    let cleanup_action = async move {
+        if !read_only {
+            let upper = fetch_upper(&mut write_handle).await;
+            let updates: Vec<_> = migration_shards_to_finalize
+                .into_iter()
+                .map(|(table_key, shard_id)| ((table_key, shard_id), upper, -1))
+                .collect();
+            // Ignore any errors, these shards will get cleaned up in the next upgrade.
```

Comment:
Probably want to `error!` it to Sentry though.

Reply:
Done